server/broadcast.go

```diff
@@ -97,7 +97,13 @@ func (bsm *BroadcastSessionsManager) completeSession(sess *BroadcastSession) {
 	bsm.sessLock.Lock()
 	defer bsm.sessLock.Unlock()
 
-	if _, ok := bsm.sessMap[sess.OrchestratorInfo.Transcoder]; ok {
+	if existingSess, ok := bsm.sessMap[sess.OrchestratorInfo.Transcoder]; ok {
+		// If the new session and the existing session share the same key in sessMap replace
+		// the existing session with the new session
+		if existingSess != sess {
+			bsm.sessMap[sess.OrchestratorInfo.Transcoder] = sess
```

Comment:
This would mutate the existing session pointer tracked by `BroadcastSessionsManager`, but the reason we call `completeSession()` with a [copy of the current session](https://github.com/livepeer/go-livepeer/pull/1254/files#diff-cecc27bfe2b6b19181ce443733dfba57R349) is to avoid mutating sessions.

Reply:
As Nico indirectly brought up, it's easy to accidentally mutate, because there's nothing explicitly preventing that.

At some point we might want to make it harder to mutate, perhaps by gating the list via an accessor, and returning a copy. This has the potential to be kind of expensive though.

It's really too bad golang doesn't have built-in immutability; the compiler and tooling could help the ergonomics, and that would facilitate a copy-on-write approach, rather than having to preemptively copy on every access. Possible that we're fighting against the grain here in terms of golang design patterns, also open to other ideas to make this more ergonomic while preserving safety.